### PR TITLE
metadb can now handle request/reply

### PR DESCRIPTION
**Pull request by divyamadala30**
Signed-off-by: Divya Madala <71040191+divyamadala30@users.noreply.github.com>

### CODE REVIEW

File: service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        final Phaser newSamplePhaser = new Phaser();
+        newSamplePhaser.register();
+        for (int lc = 0; lc < NUM_NEW_REQUEST_HANDLERS; lc++) {
+            newSamplePhaser.register();
+            exec.execute(new NewRequestReplyHandler(newSamplePhaser));
+        }
+        newSamplePhaser.arriveAndAwaitAdvance();
+    }
+    
+}
```
Comment (divyamadala30) [inline service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java:162]: Need to add an extra line here

File: service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.nats.client.Message;
+
+@Component
+public class RequestReplyHandlingServiceImpl implements RequestReplyHandlingService {
+    
+    @Value("${request_reply.patient_samples_topic}")
```
Comment (ao508) [inline service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java:27]: don't forget to add this new prop to the example properties

File: service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    ReplyInfo replyInfo = newRequestReplyQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if ((replyInfo != null)) {
+                        System.out.println("Publishing to: " + replyInfo.getReplyTo() + "\n message: "+ replyInfo.getReplyMessage());
```
Comment (n1zea144) [inline service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java:82]: Lets get rid of this output.  If we want it, we should put into a log file.

File: service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+import org.mskcc.cmo.metadb.service.RequestReplyHandlingService;
+import org.mskcc.cmo.metadb.service.SampleService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.nats.client.Message;
+
+@Component
+public class RequestReplyHandlingServiceImpl implements RequestReplyHandlingService {
```
Comment (n1zea144) [inline service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java:25]: Is the idea that this class will contain all request-reply handlers and MessagingHandlingService will contain all other handlers?

File: service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    
+    @Value("${request_reply.patient_samples_topic}")
+    private String REQUEST_REPLY_TOPIC;
```
Comment (n1zea144) [inline service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java:28]: Suggest this be called something more specific to the actual topic in the event a new request-reply topic is introduced in the future.

File: service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java (hunk 3 above)
Comment (n1zea144) [inline service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java:82]: Lets get rid of the output to stdout. If we want it, we should put into a log file.

File: service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        }        
+    }
+
+    @Override
+    public void newPatientSamplesHandler(String patientId, String replyTo) throws Exception {
+        if (!initialized) {
+            throw new IllegalStateException("Message Handling Service has not been initialized");
+        }
+        if (!shutdownInitiated) {
+            List<SampleMetadata> sampleList = sampleService.getSampleMetadataListByCmoPatientId(patientId);
```
Comment (n1zea144) [inline service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java:117]: I would move this into NewRequestReplyHandler.run() so we don't tie up the message handling thread.

**Review by n1zea144 — CHANGES_REQUESTED** [review]

**Review by n1zea144 — APPROVED** [review]

**Pull request merged**